Support for bokeh plot

techie-subhadeep opened the pull request:
This pull request adds support for creating interactive visualizations using [Bokeh](https://docs.bokeh.org/en/latest/index.html). The integration includes utilities for generating html plots and embedding them into the application.

wwwillchen commented on mesop/components/bokeh_plot/bokeh_plot.py:
```diff
@@ -0,0 +1,29 @@
+from io import BytesIO
+from typing import Protocol
+from bokeh.plotting import figure
```
can we remove this? it looks like it's not needed because figure is redefined below.

wwwillchen commented on mesop/components/bokeh_plot/bokeh_plot.py:
```diff
@@ -0,0 +1,29 @@
+from bokeh.embed import file_html
+
+
+class figure(Protocol):
```
figure -> Figure

techie-subhadeep replied:
The class name in bokeh library starts with small case [figure class github](https://github.com/bokeh/bokeh/blob/ce4ab47a86197e3742d82cbe7cb4565f840a7210/src/bokeh/plotting/_figure.py#L94)
do you still recommends the change?

wwwillchen replied:
ah, I missed that. in that case let's keep it as-is, thanks.

wwwillchen commented on mesop/components/bokeh_plot/bokeh_plot.py:
```diff
@@ -0,0 +1,29 @@
+from bokeh.plotting import figure
+from mesop.component_helpers import Style, component
+from mesop.components.html.html import html
+from bokeh.embed import file_html
```
Rather than importing here, which will make everyone using Mesop depend on bokeh, could you import bokeh  inside the bokeh_plot function? (i.e. L29).

wwwillchen reviewed (commented):
Thanks for contributing!

Could you add an E2E test? 
https://google.github.io/mesop/internal/testing/#e2e-tests

You can see some other tests here:
https://github.com/google/mesop/tree/main/mesop/tests/e2e

wwwillchen commented:
Looks like some formatting failures, could you follow: https://google.github.io/mesop/internal/development/#commit-hooks - the pre-commit hooks should fix the formatting issues. thanks.

techie-subhadeep commented:
Hello @wwwillchen 
I have fixed pre-commit lint errors
![image](https://github.com/google/mesop/assets/29140664/35d23931-d801-44fe-8e88-2ebaecb01408)

Also added e2e tests, however I am not able to run pywright tests it failed with error `error Your lockfile needs to be updated, but yarn was run with --frozen-lockfile.`

Can you please run CI again?

wwwillchen commented:
> Hello @wwwillchen I have fixed pre-commit lint errors ![image](https://private-user-images.githubusercontent.com/29140664/339498920-35d23931-d801-44fe-8e88-2ebaecb01408.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzMTYyMzgsIm5iZiI6MTcxODMxNTkzOCwicGF0aCI6Ii8yOTE0MDY2NC8zMzk0OTg5MjAtMzVkMjM5MzEtZDgwMS00NGZlLThlODgtMmViYWVjYjAxNDA4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDIxNTg1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxYjEyZGE2ODY1MjJiMjM3ZThjYTliNzY3ZGFlY2JjMGEzOGVkYTg3ZTY2ZDI1NjgwN2NkMTYzNjQ0NDQ1YWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PP9Gk64DrP5bJeT5izswOd8HgWrLNHaHmXOSlmOpbpw)
> 
> Also added e2e tests, however I am not able to run pywright tests it failed with error `error Your lockfile needs to be updated, but yarn was run with --frozen-lockfile.`
> 
> Can you please run CI again?

Re-running CI. Sometimes the npm modules gets in a weird state - try running: `bazel clean --expunge && rm -rf node_modules` and see if that fixes the yarn error.

wwwillchen commented:
I've been thinking about this some more and I think it may be better to just have bokeh be an example in the demo gallery rather than its own component because the implementation is fairly simple:

```py
  from bokeh.embed import file_html

  html(file_html(figure, "inline"), style=style)
```

Arguably the existing `plot` component (for matplotlib) shouldn't be a component and I think it was kind of borderline then, but matplotlib is the most popular Python package for plotting: https://pypistats.org/packages/matplotlib, so that was my original rationale for including it in Mesop.

ChemicalFC commented:
This approach has not been working for me, also not with plotly's own generated html file.
I do not get anything displayed (see image below).

![Screenshot 2024-07-04 113341](https://github.com/google/mesop/assets/150007456/390bfff8-4df5-4381-b9ab-d8bf5b606b5c)

**Would that issue be solved by the (already done in code) addition of "sandboxed" mode in the html component (hopefully coming soon :))?**

> I've been thinking about this some more and I think it may be better to just have bokeh be an example in the demo gallery rather than its own component because the implementation is fairly simple:
> 
> ```python
>   from bokeh.embed import file_html
> 
>   html(file_html(figure, "inline"), style=style)
> ```
> 
> Arguably the existing `plot` component (for matplotlib) shouldn't be a component and I think it was kind of borderline then, but matplotlib is the most popular Python package for plotting: https://pypistats.org/packages/matplotlib, so that was my original rationale for including it in Mesop.

wwwillchen commented:
> This approach has not been working for me, also not with plotly's own generated html file. Both make heavy use of Javascript. I do not get anything displayed (see image below). I am assuming that this has to do with the javascript since i cannot see the javascript when looking at the page using dev tools in chrome.
> 
> ![Screenshot 2024-07-04 113341](https://private-user-images.githubusercontent.com/150007456/345777689-390bfff8-4df5-4381-b9ab-d8bf5b606b5c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA4NTE5NTAsIm5iZiI6MTcyMDg1MTY1MCwicGF0aCI6Ii8xNTAwMDc0NTYvMzQ1Nzc3Njg5LTM5MGJmZmY4LTRkZjUtNDM4MS1iOWFiLWQ4YmY1YjYwNmI1Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxM1QwNjIwNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NTE5YzYwZDdhMWQxY2ZjYmU1MjZjMGM4YjQzNjVjNGM1NDcxYjUwMWZmZDA4N2RlNWVmMzJjOTNjM2FiMjEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.SyARznvNINBrDbSl59cbNy__jnvz-HpZQEXATJNBq1g)
> 
> This is the code that ends up being displayed:
> 
> ```
> <div style="background: lightgrey; border-width: 0px; border-style: initial; border-color: initial; height: 33%; width: 100%;">Bokeh Application
> <div></div>
> </div>
> ```
> 
> **Would that issue be solved by the (already done in code) addition of "sandboxed" mode in the html component (hopefully coming soon :))?**
> 
> > I've been thinking about this some more and I think it may be better to just have bokeh be an example in the demo gallery rather than its own component because the implementation is fairly simple:
> > ```python
> >   from bokeh.embed import file_html
> > 
> >   html(file_html(figure, "inline"), style=style)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Arguably the existing `plot` component (for matplotlib) shouldn't be a component and I think it was kind of borderline then, but matplotlib is the most popular Python package for plotting: https://pypistats.org/packages/matplotlib, so that was my original rationale for including it in Mesop.

hey @ChemicalFC we released [v0.9.4](https://github.com/google/mesop/releases/tag/v0.9.4) yesterday which includes the HTML sandboxed mode - please give that a shot! If you run into more issues, please feel free to open a new issue. I'll close this PR for now.

ChemicalFC commented:
Tested and worked with bokeh and plotly (via html code)
